File: netDxf/DxfDocument.MultiLeader.cs
// Copyright (c) netDxf contributors. Licensed under the MIT License.
using System.Collections.Generic;
using System.Linq;
using netDxf.Blocks;
using netDxf.Entities;
using netDxf.Objects;

namespace netDxf
{
    public sealed partial class DxfDocument
    {
        internal List<DxfObjectReference> MLeaderReferences(DxfObject target)
        {
            var result=new List<DxfObjectReference>();
            if(target is Block block)target=block.Record;
            if(target==null)return result;
            foreach(DxfObject item in this.AddedObjects.Values)
            {
                IEnumerable<DxfObject> references;
                if(item is Polyline3DRecord polylineRecord)references=polylineRecord.References;
                else if(item is DxfOpaqueEntity opaqueEntity)references=opaqueEntity.References;
                else if(item is DxfStoredSunStudy study)references=study.References;
                else if(item is DxfOpaqueObject opaqueStudy && opaqueStudy.CodeName=="SUNSTUDY")
                    references=opaqueStudy.Tags.Where(DxfObjectDatabase.IsReference).Select(tag=>this.StoredTableHandleTarget((string)tag.Value)).Where(value=>value!=null);
                else if(item is PolygonMeshRecord meshRecord)references=meshRecord.References;
                else if(item is PolyfaceMeshRecord polyfaceRecord)references=polyfaceRecord.References;
                else if(item is Polyline2DRecord legacyRecord)references=legacyRecord.References;
                else if(item is DxfStoredField field)references=field.References;
                else if(item is StoredTable table)references=table.References;
                else if(item is Section section)references=section.GeometrySettings==null?new DxfObject[0]:new DxfObject[]{section.GeometrySettings};
                else if(item is DxfSectionSettings settings)references=settings.DatabaseReferences;
                else if(item is MultiLeader leader)references=leader.Data.SelectMany(d=>d.References);
                else if(item is DxfMLeaderStyle style)references=style.DatabaseReferences;
                else if(item is DxfStoredTableContent content)references=content.References;
                else if(item is DxfStoredTableGeometry geometry)references=geometry.References;
                else if(item is DxfStoredCellStyleMap map)references=map.References;
                else if(item is DxfOpaqueObject opaqueContent && (opaqueContent.CodeName=="TABLECONTENT" || opaqueContent.CodeName=="TABLEGEOMETRY" || opaqueContent.CodeName=="CELLSTYLEMAP"))
                    references=opaqueContent.Tags.Where(DxfObjectDatabase.IsReference).Select(tag=>this.StoredTableHandleTarget((string)tag.Value)).Where(value=>value!=null);
                else if(item is DxfTableStyle tableStyle)references=tableStyle.References;
                else if(item is DxfStoredDimAssoc association)references=association.References;
                else if(item is DxfOpaqueObject opaque && opaque.CodeName=="DIMASSOC")references=opaque.Tags.Where(DxfObjectDatabase.IsReference).Select(tag=>this.StoredTableHandleTarget((string)tag.Value));
                else continue;
                int count=references.Count(r=>ReferenceEquals(r,target));
                if(count>0)result.Add(new DxfObjectReference(item,count));
            }
            return result;
        }
    }
}

File: javascript/netDxf/DxfDocument.MultiLeader.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import * as api from '../index.js';
import { ReferenceList } from '../runtime/ReferenceList.js';
const is = (item, name) => typeof api[name] === 'function' && item instanceof api[name];
/** Live retained-reference accounting. Counts repeated references by identity, not names. */
export function InstallDocumentMultiLeader(Type) {
  Type.prototype.StoredTableHandleTarget = function(handle) {
    const registered = this.GetObjectByHandle(handle); if (registered !== null) return registered;
    if (typeof handle !== 'string' || !/^[0-9a-f]{1,16}$/i.test(handle) || BigInt('0x' + handle) === 0n) return null;
    const value = BigInt('0x' + handle);
    for (const item of this.RetainedMetadataObjects())
      if (typeof item.Handle === 'string' && /^[0-9a-f]{1,16}$/i.test(item.Handle) && BigInt('0x' + item.Handle) === value) return item;
    return null;
  };
  Type.prototype.MLeaderReferences = function(target) {
    const result = new ReferenceList(); if (target instanceof api.Block) target = target.Record;
    if (target == null) return result;
    for (const item of this.AddedObjects.Values) {
      let references;
      if (['Polyline3DRecord','DxfOpaqueEntity','DxfStoredSunStudy','PolygonMeshRecord','PolyfaceMeshRecord',
           'Polyline2DRecord','DxfStoredField','StoredTable','DxfStoredTableContent','DxfStoredTableGeometry',
           'DxfStoredCellStyleMap','DxfTableStyle','DxfStoredDimAssoc'].some(name => is(item,name))) references = item.References;
      else if (item instanceof api.Section) references = item.GeometrySettings === null ? [] : [item.GeometrySettings];
      else if (item instanceof api.DxfSectionSettings || item instanceof api.DxfMLeaderStyle) references = item.DatabaseReferences;
      else if (item instanceof api.MultiLeader) references = Array.from(item.Data).flatMap(data => Array.from(data.References));
      else if (item instanceof api.DxfOpaqueObject && ['SUNSTUDY','TABLECONTENT','TABLEGEOMETRY','CELLSTYLEMAP','DIMASSOC'].includes(item.CodeName))
        references = Array.from(item.Tags).filter(api.DxfObjectDatabase.IsReference).map(tag => this.StoredTableHandleTarget(tag.Value));
      else continue;
      let uses = 0; for (const reference of references) if (reference === target) uses++;
      if (uses > 0) result.Add(new api.DxfObjectReference(item,uses));
    }
    return result;
  };
}
